node/node.go

```diff
@@ -701,3 +708,73 @@ func (node *Node) ShutDown() {
 	fmt.Print(msg)
 	os.Exit(0)
 }
+
+func (node *Node) populateSelfAddresses(epoch *big.Int) {
```

Comment:
Also, the mapping from BLS-> ecdsa address should never change. a BLS is always tied to the first registered validator. So it's safe to just set the mapping once at start.

Reply:
Or we could just maintain the map without reset and keep updating it when new epoch changes.

Reply:
it is a valid scenario to remove bls key from one validator and add it to another? especially when a staking partner has many validators?

Reply:
ok, you are right. Then let's make this part thread safe. Make sure it doesn't crash

Reply:
since this map is only accessed linearly, will keep it thread unsafe. It is accessed by `proposeNewBlock` and `PostConsensusProcessing`.